src/wallet/rpcwallet.cpp

```diff
@@ -2737,7 +2737,7 @@ static UniValue createwallet(const JSONRPCRequest& request)
 
     UniValue obj(UniValue::VOBJ);
     obj.pushKV("name", wallet->GetName());
-    obj.pushKV("warning", Join(warnings, "\n", OpOriginal));
+    obj.pushKV("warning", Join(warnings, std::string("\n"), OpOriginal));
```

Comment:
Please keep the explicit overload for std::string, if you want you can extend it with a comment why it exists.

Also, I like the idea of deriving the return type from the unary operator

```diff
diff --git a/src/util/string.h b/src/util/string.h
index 81d4c186e5..e7651c3e7d 100644
--- a/src/util/string.h
+++ b/src/util/string.h
@@ -31,9 +31,10 @@ NODISCARD inline std::string TrimString(const std::string& str, const std::strin
  * @param unary_op   Apply this operator to each item in the list
  */
 template <typename T, typename BaseType, typename UnaryOp>
-BaseType Join(const std::vector<T>& list, const BaseType& separator, UnaryOp unary_op)
+auto Join(const std::vector<T>& list, const BaseType& separator, UnaryOp unary_op)
+    -> decltype(unary_op(list.at(0)))
 {
-    BaseType ret;
+    decltype(unary_op(list.at(0))) ret;
     for (size_t i = 0; i < list.size(); ++i) {
         if (i > 0) ret += separator;
         ret += unary_op(list.at(i));
@@ -47,6 +48,12 @@ T Join(const std::vector<T>& list, const T& separator)
     return Join(list, separator, [](const T& i) { return i; });
 }
 
+inline std::string Join(const std::vector<std::string>& list, const std::string& separator)
+{
+    // explicit overload needed for c_str, which would otherwise cause a substitution failure in the template above
+    return Join<std::string>(list, separator);
+}
+
 /**
  * Check if a string does not contain any embedded NUL (\0) characters
  */

Reply:
> Please keep the explicit overload for std::string, if you want you can extend it with a comment why it exists.

Is it just to avoid writing `std::string("\n")` instead of `"\n"` in the `Join()` argument list?

Reply:
[Updated](https://github.com/bitcoin/bitcoin/pull/18922#issuecomment-626273821).